File: sections/35_math_mixed_contest/problems/q_coprime_rectangles/solution.cpp
#include <bits/stdc++.h>
using namespace std;

const long long MOD = 1'000'000'007;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int q;
    cin >> q;
    vector<pair<int, int>> queries(q);
    int maximum = 1;
    for (auto& [a, b] : queries) {
        cin >> a >> b;
        maximum = max(maximum, min(a, b));
    }

    vector<int> mu(maximum + 1), primes;
    vector<char> composite(maximum + 1);
    mu[1] = 1;
    for (int i = 2; i <= maximum; ++i) {
        if (!composite[i]) {
            primes.push_back(i);
            mu[i] = -1;
        }
        for (int p : primes) {
            if (1LL * i * p > maximum)
                break;
            composite[i * p] = true;
            if (i % p == 0)
                mu[i * p] = 0;
            else
                mu[i * p] = -mu[i];
            if (i % p == 0)
                break;
        }
    }

    vector<int> prefix(maximum + 1);
    for (int i = 1; i <= maximum; ++i)
        prefix[i] = prefix[i - 1] + mu[i];

    for (auto [a, b] : queries) {
        long long answer = 0;
        int left = 1;
        int limit = min(a, b);
        while (left <= limit) {
            int qa = a / left;
            int qb = b / left;
            int right = min(limit, min(a / qa, b / qb));
            long long coefficient = prefix[right] - prefix[left - 1];
            answer = (answer + coefficient * qa % MOD * qb) % MOD;
            left = right + 1;
        }
        cout << (answer + MOD) % MOD << '\n';
    }
}

File: sections/35_math_mixed_contest/problems/q_coprime_rectangles/solution.py
import sys

MOD = 1_000_000_007


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    q = data[0]
    queries = [(data[i], data[i + 1]) for i in range(1, 2 * q + 1, 2)]
    maximum = max((min(a, b) for a, b in queries), default=1)

    mu = [0] * (maximum + 1)
    primes = []
    composite = [False] * (maximum + 1)
    mu[1] = 1
    for value in range(2, maximum + 1):
        if not composite[value]:
            primes.append(value)
            mu[value] = -1
        for prime in primes:
            product = value * prime
            if product > maximum:
                break
            composite[product] = True
            if value % prime == 0:
                mu[product] = 0
                break
            mu[product] = -mu[value]

    prefix = [0] * (maximum + 1)
    for value in range(1, maximum + 1):
        prefix[value] = prefix[value - 1] + mu[value]

    answers = []
    for a, b in queries:
        answer = 0
        left = 1
        limit = min(a, b)
        while left <= limit:
            qa = a // left
            qb = b // left
            right = min(limit, a // qa, b // qb)
            answer += (prefix[right] - prefix[left - 1]) * qa * qb
            left = right + 1
        answers.append(str(answer % MOD))
    print("\n".join(answers))


if __name__ == "__main__":
    main()
